frontend/src/utils/sqlEditor.ts

```diff
@@ -206,49 +206,35 @@ export const ensureDataSourceSelection = (
     (ds) => ds.type === DataSourceType.READ_ONLY
   );
 
-  let behavior: "RO" | "FALLBACK" | "ALLOW_ADMIN";
+  let fallback = false;
+  // Fallback to the admin data source is permitted if neither the environment policy
+  // nor the project policy is set to DISALLOW, and there is no read-only (RO) data source available.
   if (
-    restriction.environmentPolicy ===
-      DataSourceQueryPolicy_Restriction.DISALLOW ||
-    restriction.projectPolicy === DataSourceQueryPolicy_Restriction.DISALLOW
+    restriction.environmentPolicy !==
+      DataSourceQueryPolicy_Restriction.DISALLOW &&
+    restriction.projectPolicy !== DataSourceQueryPolicy_Restriction.DISALLOW
   ) {
-    behavior = "RO";
-  } else if (
-    restriction.environmentPolicy ===
-      DataSourceQueryPolicy_Restriction.FALLBACK ||
-    restriction.projectPolicy === DataSourceQueryPolicy_Restriction.FALLBACK
-  ) {
-    behavior = "FALLBACK";
-  } else {
-    behavior = "ALLOW_ADMIN";
+    fallback = true;
   }
-  if (behavior === "ALLOW_ADMIN") {
-    if (current) {
-      return current;
-    }
-    return adminDataSource.id;
+
+  if (current && readonlyDataSources.some((ds) => ds.id === current)) {
+    // If the current data source is a read-only data source, we can use it.
+    return current;
   }
-  if (behavior === "FALLBACK") {
-    if (current) {
-      return current;
-    }
-    return head(readonlyDataSources)?.id ?? adminDataSource.id;
+
+  let ds = head(readonlyDataSources);
+  if (!ds && fallback) {
+    ds = adminDataSource;
```

Comment:
current=ADMIN datasource. policy=ALLOW_ADMIN.
However, my selection gets overwritten to RO